<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>5. Analysis of Metagenomic data in R &mdash; PhD-BEE Tor Vergata - Metagenomic analysis of NGS data 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="6. Do it yourself" href="6_Do_it_yourself.html" />
    <link rel="prev" title="4. Prepare abundance tables for downstream analyses" href="4_prepare_tables.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> PhD-BEE Tor Vergata - Metagenomic analysis of NGS data
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. Course material</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_ReadsFiltering.html">2. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Metagenomics_taxonomy.html">3. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_prepare_tables.html">4. Prepare abundance tables for downstream analyses</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">5. Analysis of Metagenomic data in R</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#install-r-packages">5.1. Install R packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#importing-abundance-tables-in-phyloseq">5.2. Importing abundance tables in Phyloseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="#barplot-of-taxa-abundances">5.3. Barplot of taxa abundances</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alpha-diversity">5.4. Alpha diversity</a></li>
<li class="toctree-l2"><a class="reference internal" href="#note-on-compositional-data">5.5. Note on compositional data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#principal-coordinate-analysis-multidimensional-scaling">5.6. Principal Coordinate Analysis (Multidimensional Scaling)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#non-metric-multidimensional-scaling">5.7. Non-Metric Multidimensional Scaling</a></li>
<li class="toctree-l2"><a class="reference internal" href="#differential-taxonomic-abundances-with-deseq2">5.8. Differential taxonomic abundances with DESeq2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#principal-component-analysis-pca">5.9. Principal Component Analysis (PCA)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="6_Do_it_yourself.html">6. Do it yourself</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PhD-BEE Tor Vergata - Metagenomic analysis of NGS data</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">5. </span>Analysis of Metagenomic data in R</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/5_Metagenomic_analysis_R.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="analysis-of-metagenomic-data-in-r">
<h1><span class="section-number">5. </span>Analysis of Metagenomic data in R<a class="headerlink" href="#analysis-of-metagenomic-data-in-r" title="Permalink to this headline"></a></h1>
<p>Now that we have formatted out abundance table we can import it in R.</p>
<section id="install-r-packages">
<h2><span class="section-number">5.1. </span>Install R packages<a class="headerlink" href="#install-r-packages" title="Permalink to this headline"></a></h2>
<p>We will use the R package <code class="docutils literal notranslate"><span class="pre">Phyloseq</span></code> to explore the microbiome data in the normalized abundance table. Once imported in R, the table can be tweaked to focus on a subset of samples.
Install Phyloseq from <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html">Bioconductor</a>.</p>
<blockquote>
<div></div></blockquote>
<p>Let’s first set up the work-directory containing the table that we generated:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;path/to/folder&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>If needed, install the following R packages:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;phyloseq&quot;</span><span class="p">)</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;ggplot2&quot;</span><span class="p">)</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;vegan&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The packages <strong>microbiome</strong> and <strong>DESeq2</strong> must be installed via <strong>BiocManager</strong></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>

<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;microbiome&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="importing-abundance-tables-in-phyloseq">
<h2><span class="section-number">5.2. </span>Importing abundance tables in Phyloseq<a class="headerlink" href="#importing-abundance-tables-in-phyloseq" title="Permalink to this headline"></a></h2>
<p>Phyloseq works with <code class="docutils literal notranslate"><span class="pre">biom</span></code> objects that store all the information required for the analyses: abundance of taxa, sample metadata, taxonomy
First you can import the table with the species abundances and the full taxonomic ranks. This table has the taxa in the rows and the samples in the columns.
Note that we will set the column #20, corresponding to the species names, as row names.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">table</span> <span class="o">=</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;abundance_table_IDs.merged.norm.taxonomy.final.Archaea_Bacteria&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>From this table, we will extract as separate objects the taxonomy and the species abundances.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract the taxonomy by selecting the corresponding columns.</span>
<span class="n">taxonomy</span> <span class="o">=</span> <span class="p">(</span><span class="n">table</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="o">:</span><span class="m">19</span><span class="p">)])</span>
<span class="c1"># extract the otu by removing the columns corresponding to the taxonomy and other unwanted columns.</span>
<span class="n">otu</span> <span class="o">=</span> <span class="p">(</span><span class="n">table</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">23</span><span class="p">)])</span>
<span class="c1"># make the matrix values as numeric</span>
<span class="nf">class</span><span class="p">(</span><span class="n">otu</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;numeric&quot;</span>
</pre></div>
</div>
<p>Then you can generate the <code class="docutils literal notranslate"><span class="pre">biom</span></code> object with phyloseq.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">phyloseq</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">vegan</span><span class="p">)</span>
<span class="c1"># assing otu and tax tables</span>
<span class="n">OTU</span> <span class="o">=</span> <span class="nf">otu_table</span><span class="p">(</span><span class="n">otu</span><span class="p">,</span> <span class="n">taxa_are_rows</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">TAX</span> <span class="o">=</span> <span class="nf">tax_table</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">)</span>
<span class="n">my_biom</span> <span class="o">=</span> <span class="nf">phyloseq</span><span class="p">(</span><span class="n">OTU</span><span class="p">,</span> <span class="n">TAX</span><span class="p">)</span>
</pre></div>
</div>
<p>As additional step, you can add to the biom file the metadata associated with the samples. These will be useful to describe the samples and select them in the downstream analyses (e.g. to work on subset of samples).
The metadata are stored in a mapping file, a tab-delimited text file where you have in the first column the names of your samples (matching the names in the columns of your OTU abundance table), and in the other columns as many metadata as you want (e.g. sample source, literature ID etc.).
You can download the metadata of the dataset that we are analysing: <a class="reference external" href="https://drive.google.com/file/d/1oedAI5qA_0RsVEL20oe8Yf80WmGnMzYR/view?usp=sharing">metadata_chimps.txt</a></p>
<blockquote>
<div></div></blockquote>
<p>The metadata can be incorporated in your biom file as follows:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># import mapping file with metadata</span>
<span class="n">biom_metadata</span> <span class="o">&lt;-</span> <span class="nf">import_qiime_sample_data</span><span class="p">(</span><span class="s">&quot;metadata_chimps.txt&quot;</span><span class="p">)</span>
<span class="c1"># merge data</span>
<span class="n">my_biom</span> <span class="o">&lt;-</span> <span class="nf">merge_phyloseq</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">biom_metadata</span><span class="p">)</span>
</pre></div>
</div>
<p>Once the biom object has been generated, you can try some commands to explore it:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># check sample names:</span>
<span class="nf">sample_data</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="c1"># check taxa:</span>
<span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="c1"># check rank names:</span>
<span class="nf">colnames</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If needed, it is convenient to rename the taxa columns more appropriately (some tools to retrieve the full taxonomy may return “Rank1” etc.)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># check the labels used for rank names</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span>
<span class="c1"># change them if needed</span>
<span class="nf">colnames</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Kingdom&quot;</span><span class="p">,</span> <span class="s">&quot;Phylum&quot;</span><span class="p">,</span> <span class="s">&quot;Class&quot;</span><span class="p">,</span> <span class="s">&quot;Order&quot;</span><span class="p">,</span> <span class="s">&quot;Family&quot;</span><span class="p">,</span> <span class="s">&quot;Genus&quot;</span><span class="p">,</span> <span class="s">&quot;Species&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>In the next step we’ll remove low-abundance taxa below a desired threshold (most commonly 0.02% is chosen).
This is done because the very rare taxa (below the threshold) may be false positive resulting from taxonomic misassignments.
To do that we will create various objects and use a function to remove the taxa below the desired threshold.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">minTotRelAbun</span> <span class="o">=</span> <span class="m">0.0002</span>
<span class="n">x</span> <span class="o">=</span> <span class="nf">taxa_sums</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="n">keepTaxa</span> <span class="o">=</span> <span class="nf">taxa_names</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)[</span><span class="nf">which</span><span class="p">((</span><span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">minTotRelAbun</span><span class="p">)]</span>
<span class="n">my_biom_flt</span> <span class="o">=</span> <span class="nf">prune_taxa</span><span class="p">(</span><span class="n">keepTaxa</span><span class="p">,</span> <span class="n">my_biom</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, we convert the absolute abundance values in relative abundance. This works as normalization for the different sequencing depths of the datasets that we are analysing.
To do that we use the function <code class="docutils literal notranslate"><span class="pre">transform_sample_counts</span></code> in Phyloseq</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">my_biom_rel</span> <span class="o">=</span> <span class="nf">transform_sample_counts</span><span class="p">(</span><span class="n">my_biom_flt</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="barplot-of-taxa-abundances">
<h2><span class="section-number">5.3. </span>Barplot of taxa abundances<a class="headerlink" href="#barplot-of-taxa-abundances" title="Permalink to this headline"></a></h2>
<p>Now that our table is formatted as bion object, we can start to analyse it by generating barplots to display the relative abundance of each sample at different taxonomic ranks.
To do that we will first merge the abundance of the taxa of our table (e.g. species) to a higher taxonomic rank (e.g. Phylum).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge taxa at the Phylum rank with tax_glom (in Phyloseq)</span>
<span class="n">my_biom_phylum</span> <span class="o">=</span> <span class="nf">tax_glom</span><span class="p">(</span><span class="n">my_biom_flt</span><span class="p">,</span> <span class="s">&quot;phylum&quot;</span><span class="p">,</span> <span class="n">NArm</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="c1"># convert to relative freq</span>
<span class="n">my_biom_phylum_rel</span> <span class="o">=</span> <span class="nf">transform_sample_counts</span><span class="p">(</span><span class="n">my_biom_phylum</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
<span class="c1"># plot the relative abundances (you can use ggplot2 themes to improve the chart)</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">my_biom_phylum_rel</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;phylum&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>In the <code class="docutils literal notranslate"><span class="pre">tax_glom</span></code> command the label of the rank must match the one that is defined in your original biom file, as reported in the column names of the tax_table command (see above).</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You may see that in some samples Kraken2 did not return any taxonomic classification. This is because the database is too small (2Gb!), and in some instance there may not enough resolution to make the taxonomic classification.
Identify the sample (e.g. from the Phyla barplot, remove it from the original table in R and create again the biom file).
You can see the difference in the results by repeating the analyses on the same dataset (including more chimps) generated by using a Kraken2 database of 8Gb (see next secion).</p>
</div>
<p>Try to generate barplots at different taxonomic ranks (e.g. family).
To save the plot in the pdf format you can use this command:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">dev.print</span><span class="p">(</span><span class="n">pdf</span><span class="p">,</span> <span class="s">&#39;filename.pdf)&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also focus your analysis on a subset of samples and use the metadata contained in your original mapping file. It would be convenient to add metadata the way it more convenient to select and compare your samples. Here we subsample for some groups of individuals as defined in the metadata column <strong>Group1</strong>.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample single group phyla</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel_phylum</span><span class="p">,</span> <span class="n">Group1</span><span class="o">==</span><span class="s">&quot;Dental calculus chimp&quot;</span><span class="p">)</span>
<span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Dental_calculus_chimp&quot;</span><span class="p">,</span><span class="s">&quot;Dental_calculus&quot;</span><span class="p">,</span><span class="s">&quot;Oral_plaque&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel_phylum</span><span class="p">,</span> <span class="n">Group1</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1"># barplots in groups with facet_grid</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s">&quot;phylum&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">facet_grid</span><span class="p">(</span><span class="o">~</span><span class="n">Group1</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span> <span class="s">&quot;free_x&quot;</span><span class="p">,</span> <span class="n">space</span> <span class="o">=</span> <span class="s">&quot;free_x&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">6</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
<p>Try now to generate barplots at the family level. You will end-up with more labels in the legend, to tweak it play with the options <code class="docutils literal notranslate"><span class="pre">legend.position=&quot;right&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">legend.text</span> <span class="pre">=</span> <span class="pre">element_text(size</span> <span class="pre">=</span> <span class="pre">4)</span></code>, <code class="docutils literal notranslate"><span class="pre">legend.key.size</span> <span class="pre">=</span> <span class="pre">unit(0.5,&quot;line&quot;)</span></code>, inside <code class="docutils literal notranslate"><span class="pre">theme</span></code> in ggplot2.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_bar</span><span class="p">(</span><span class="n">my_biom_family_rel</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;family&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;right&quot;</span><span class="p">,</span> <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">4</span><span class="p">),</span> <span class="n">legend.key.size</span> <span class="o">=</span> <span class="nf">unit</span><span class="p">(</span><span class="m">0.5</span><span class="p">,</span><span class="s">&quot;line&quot;</span><span class="p">),</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="alpha-diversity">
<h2><span class="section-number">5.4. </span>Alpha diversity<a class="headerlink" href="#alpha-diversity" title="Permalink to this headline"></a></h2>
<p>Alpha diversity (α-diversity) is defined as the mean diversity of species in different sites or habitats within a local scale, it is used to describe the “within-sample” diversity.
A number of metrics are commonly-used to measure alpha diversity, such as the Shannon index and the Simpson index (read more about <a class="reference external" href="https://docs.onecodex.com/en/articles/4136553-alpha-diversity">Alpha diversity</a>).
It is advisable to use untrimmed, non-normalized count data for meaningful results, as many of alpha diversity estimates are highly dependent on the number of singletons (taxa with count=1). You can always trim the data later on if needed, just not before using this function.</p>
<blockquote>
<div></div></blockquote>
<p>We will estimate the Shannon index and the Simpson index on our dataset and plot them with the function <code class="docutils literal notranslate"><span class="pre">plot_richness</span></code> of Phyloseq.
For alpha diversity estimation, Phyloseq accepts only integer values, so first we’ll round out table to integers and then plot the alpha diversity values for each sample and for group of samples (as identified in the Group1 metadata).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># make the data in the otu table integer</span>
<span class="nf">otu_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span>
<span class="c1"># plot two diversity indices as dots for each individual and customize legend and axis-x</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">measures</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Simpson&quot;</span><span class="p">,</span> <span class="s">&quot;Shannon&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
      <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;right&quot;</span><span class="p">,</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
<span class="c1"># plot richness based on group variable</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span> <span class="n">measures</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Simpson&quot;</span><span class="p">,</span> <span class="s">&quot;Shannon&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span>
<span class="c1"># plot richness as violin plot</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span> <span class="n">measures</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Simpson&quot;</span><span class="p">,</span> <span class="s">&quot;Shannon&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
      <span class="nf">geom_violin</span><span class="p">()</span>
<span class="c1"># plot richness by including fots in the violin plots</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span> <span class="n">measures</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Simpson&quot;</span><span class="p">,</span> <span class="s">&quot;Shannon&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
      <span class="nf">geom_violin</span><span class="p">()</span> <span class="o">+</span>
      <span class="nf">geom_dotplot</span><span class="p">(</span><span class="n">binaxis</span><span class="o">=</span><span class="s">&#39;y&#39;</span><span class="p">,</span> <span class="n">stackdir</span><span class="o">=</span><span class="s">&#39;center&#39;</span><span class="p">,</span> <span class="n">dotsize</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="c1"># plot richness as boxplots - you can also combine violin and boxplot</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span> <span class="n">measures</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Simpson&quot;</span><span class="p">,</span> <span class="s">&quot;Shannon&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_boxplot</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="m">0.3</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="note-on-compositional-data">
<h2><span class="section-number">5.5. </span>Note on compositional data<a class="headerlink" href="#note-on-compositional-data" title="Permalink to this headline"></a></h2>
<p>An important concept to keep in mind for microbiome analysis from NGS data is that <strong>microbiome datasets are compositional</strong>, which means that they have an arbitrary total imposed by the sequencing instrument. This is due to the fact that high-throughput sequencing (HTS) instruments can deliver reads only up to the capacity of the instrument.
Thus, the total read count observed in an NGS run is a fixed-size, random sample of the relative abundance of the molecules in the underlying ecosystem (e.g. oral microbiota).
Several methods applied include count-based strategies (normalized to a constant area or volume, e.g. the sequencing lane output) such as Bray-Curtis dissimilarity, zero-inflated Gaussian models and negative binomial models, but these do not account for the limitations imposed by the instrument and the so-called principle of true indpendence of species in ecology.
Read more about compositonal data in the paper by <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full">Gloor et al. (2017)</a> and <a class="reference external" href="https://academic.oup.com/gigascience/article/8/9/giz107/5572529">Quinn et al. (2019)</a>.</p>
<blockquote>
<div></div></blockquote>
<p>Due to the compositional nature of microbiome datasets, the centered log-ratio (clr) transformation introduced by Aitchison (1986) is often used. The clr-transformed values are scale-invariant, which means that the same ratio is expected to be obtained in a sample with few read counts or an identical sample with many read counts, only the precision of the clr estimate is affected.
The clr transformation uses the geometric mean g(x) of the sample vector as reference.
Given an observation vector of D “counted” features (taxa, operational taxonomic units or OTUs, genes, etc.) in a sample, x = [x1, x2, …xD], the geometric mean g(x) is:</p>
<div class="math notranslate nohighlight">
\[g(x) = \sqrt[D]{x_{i} * ... *  x_{D}}\]</div>
<p>The log-ratio transformation is applied to each subject vector <em>x</em> with <em>D</em> features (OTUs). Basically, the analysis of clr data will reveal how OTUs behave relative to the per-sample average.</p>
<div class="math notranslate nohighlight">
\[clr(x) = [\ln{\frac {x_{i}}{g(x)}},...,\ln{\frac {x_{D}}{g(x)}}]\]</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Compositional methods depend on logarithms that <strong>do not compute for zeros</strong>. Therefore, zeros in the abundance table must be addressed prior to, or during, the pipeline.
Two different methods of zero-handling can be used: zero-replacement using the pseudo-counts method from the R package <a class="reference external" href="https://www.sciencedirect.com/user/identity/landing?code=B0sdK2bMVyKF904thdCXKg1gPy2uFRWS5qoj4UPg&amp;state=retryCounter%3D0%26csrfToken%3Deb9aa274-f4e8-4f8c-9adc-04d3770e7d74%26idpPolicy%3Durn%253Acom%253Aelsevier%253Aidp%253Apolicy%253Aproduct%253Ainst_assoc%26returnUrl%3D%252Fscience%252Farticle%252Fpii%252FS0169743915000490%26prompt%3Dnone%26cid%3Darp-25746d25-eda8-434f-bad4-f41bcdbb30ca">zCompositions</a> version 1.3.3 and the use of an arbitrary “offset”, which means replacing the zeroes with ones.
It is good-praxis to always demonstrate that the removal or modification of zero-laden features does not change the overall interpretation of the results (<a class="reference external" href="https://academic.oup.com/gigascience/article/8/9/giz107/5572529">Quinn et al. (2019)</a>).</p>
<blockquote>
<div></div></blockquote>
</div>
<p>To account for the compositional nature of microbiome datasets, in the following analyses we will use Aitchinson-based (namely, clr-transformed) abundance values.</p>
</section>
<section id="principal-coordinate-analysis-multidimensional-scaling">
<h2><span class="section-number">5.6. </span>Principal Coordinate Analysis (Multidimensional Scaling)<a class="headerlink" href="#principal-coordinate-analysis-multidimensional-scaling" title="Permalink to this headline"></a></h2>
<p>The Principal Coordinate Analysis (PCoA), also referred to as metric Multidimensional Scaling (MDS) is a multivariate reduction method performed on distance (or dissimilarity) indexes.
In fact, PCoA can handle (dis)similarity matrices calculated from quantitative, semi-quantitative, qualitative, and mixed variables.
Here we will use the Aitchinson distance, which corresponds to the Euclidean distance of the CLR-transformed dataset of relative abundances. To make of the clr-transformation of the abundance data we will use the library <code class="docutils literal notranslate"><span class="pre">microbiome</span></code>.
Read more here about the MDS and so-called <a class="reference external" href="https://ourcodingclub.github.io/tutorials/ordination/">ordination methods</a></p>
<blockquote>
<div></div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The CLR transformation of the package microbiome automatically deals with zeroes by appliying a pseudocount of min(relative abundance)/2 to exact zero relative abundance entries in the OTU table before taking logs.</p>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculate Aitchinson distance (the Aitchison distance is a sub-compositionally coherent distance defined as the Euclidean distance after the clr-transformation of the compositions).</span>
<span class="nf">library</span><span class="p">(</span><span class="n">microbiome</span><span class="p">)</span>
<span class="n">my_biom_rel_clr</span> <span class="o">=</span> <span class="nf">transform</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="s">&quot;clr&quot;</span><span class="p">)</span>
<span class="n">distAIT</span> <span class="o">=</span> <span class="nf">distance</span><span class="p">(</span><span class="n">my_biom_rel_clr</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;euclidean&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To make the PCoA we will use the <code class="docutils literal notranslate"><span class="pre">ordinate</span></code> function in Phyloseq.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># make PCoA ordination (MDS) of Aitchison distances</span>
<span class="n">ordAIT</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">my_biom_rel_clr</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;PCoA&quot;</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">distAIT</span><span class="p">)</span>
<span class="c1"># plot PCoA (MDS)</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">my_biom_rel_clr</span><span class="p">,</span> <span class="n">ordAIT</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group1&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can also investigate how much of the total distance structure we will capture in the first few axes. We can do this graphically with a <strong>scree plot</strong>, an ordered barplot of the relative fraction of the total eigenvalues associated with each axis.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_scree</span><span class="p">(</span><span class="n">ordAIT</span><span class="p">,</span> <span class="s">&quot;Scree plot for PCoA of Aitchinson distances&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Like we already did above, you can make the MDS on a subset of samples:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Dental_calculus_chimp&quot;</span><span class="p">,</span><span class="s">&quot;Dental_calculus&quot;</span><span class="p">,</span><span class="s">&quot;Oral_plaque&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="n">Group1</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1"># calculate Aitchinson distance</span>
<span class="nf">library</span><span class="p">(</span><span class="n">microbiome</span><span class="p">)</span>
<span class="n">subsample_clr</span> <span class="o">=</span> <span class="nf">transform</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="s">&quot;clr&quot;</span><span class="p">)</span>
<span class="n">distAIT</span> <span class="o">=</span> <span class="nf">distance</span><span class="p">(</span><span class="n">subsample_clr</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;euclidean&quot;</span><span class="p">)</span>
<span class="c1"># make the PCoA ordination (MDS) setting the colors of the points automatically as defined in the metadata &quot;Group1&quot;.</span>
<span class="n">ordAIT</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;PCoA&quot;</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">distAIT</span><span class="p">)</span>
<span class="c1"># two-dimension plot PCoA (MDS) setting the colors of the points automatically as defined in the metadata &quot;Group2&quot;.</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">ordAIT</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="non-metric-multidimensional-scaling">
<h2><span class="section-number">5.7. </span>Non-Metric Multidimensional Scaling<a class="headerlink" href="#non-metric-multidimensional-scaling" title="Permalink to this headline"></a></h2>
<p>Unlike other ordination techniques that rely on (primarily Euclidean) distances, such as Principal Coordinates Analysis, nMDS uses rank orders (so not the abundances), for this reason it is non-metric.
You can read more on the nMDS in the blogs curated by ecologists: <a class="reference external" href="https://archetypalecology.wordpress.com/2018/02/18/non-metric-multidimensional-scaling-nmds-what-how/">link1</a></p>
<p>To begin, NMDS requires a matrix of dissimilarities. Raw Euclidean distances are not ideal for this purpose: they’re sensitive to total abundances.
Previously, NMDS plots based on Bray-Curtis distances were most commonly used with rarefied and relative abundance transformations, but Euclidean distances with clr-transformed data are becoming more common now. See for example <a class="reference external" href="https://academic.oup.com/nargab/article/2/4/lqaa079/5917299?login=true">Sisk-Hackworth &amp; Kelley (2020)</a>.
NMDS arranges points to maximize rank-order correlation between real-world distance and ordination space distance.</p>
<blockquote>
<div></div></blockquote>
<p>The technique uses a <em>trial and error</em> to find the best positioning in dimensional space. Goodness-of-fit is measured by <strong>stress</strong> – a measure of rank-order disagreement between observed and fitted distances.
A stress values below 0.2 ic considered acceptable.
You can follow the same steps as above, just change the ordinatiom method supported in the command.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># make PCoA ordination (MDS) of Aitchison distances</span>
<span class="n">ordAIT</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">my_biom_rel_clr</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;NMDS&quot;</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">distAIT</span><span class="p">)</span>
<span class="c1"># plot the NMDS setting the colors of the points automatically as defined in the metadata &quot;Group1&quot;.</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">my_biom_rel_clr</span><span class="p">,</span> <span class="n">ordAIT</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="differential-taxonomic-abundances-with-deseq2">
<span id="deseq2"></span><h2><span class="section-number">5.8. </span>Differential taxonomic abundances with DESeq2<a class="headerlink" href="#differential-taxonomic-abundances-with-deseq2" title="Permalink to this headline"></a></h2>
<p>A common goal of microbiome studies is to identify differentially abundant taxa (species, genera etc.) between different groups of samples.
One of the tools used to do that is DESeq2, which was originally developed to identify differentially expressed genes in RNAseq data, but it is commonly adopted also in microbiome studies. You can read more about DESeq2 in the <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">original publication</a> and in the <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">dedicated page</a> of Bioconductor.
You can install <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html">DESeq2 from Bioconductor</a>.</p>
<blockquote>
<div></div></blockquote>
<p>Run DESeq2 on the raw abundace data (<a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts">here is why</a>), filtered for low-abundance taxa. You can choose to work on a subset of sample, as done above. For example, we can select only the oral samples.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span>
<span class="c1"># subsample oral microbiomes</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Dental_calculus_chimp&quot;</span><span class="p">,</span><span class="s">&quot;Dental_calculus&quot;</span><span class="p">,</span><span class="s">&quot;Oral_plaque&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">Group1</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
</pre></div>
</div>
<p>Then, we filter out low-abundance taxa with a mean &lt; 0.02%.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">minTotRelAbun</span> <span class="o">=</span> <span class="m">0.0002</span>
<span class="n">x</span> <span class="o">=</span> <span class="nf">taxa_sums</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span>
<span class="n">keepTaxa</span> <span class="o">=</span> <span class="nf">taxa_names</span><span class="p">(</span><span class="n">subsample</span><span class="p">)[</span><span class="nf">which</span><span class="p">((</span><span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">minTotRelAbun</span><span class="p">)]</span>
<span class="n">subsample_flt</span> <span class="o">=</span> <span class="nf">prune_taxa</span><span class="p">(</span><span class="n">keepTaxa</span><span class="p">,</span> <span class="n">subsample</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is better to remove the low-abundant taxa only <strong>after</strong> subsampling the abudance table. Doing this before will keep unwanted taxa that are more common in other microbiota (e.g. soil) and irrelevant for describing oral envrionments (or the specific environment you are analysing with the PCA).</p>
</div>
<p>DESeq2 does not tolerate zero-counts, for this reason an offset (+1) is commonly applied to remove all zeroes from your table.
After that, run DESeq2 on your table and contrat two sets of samples to find differential taxa abundances.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># make a +1 offset to run Deseq2 (which does not tolerate zero counts)</span>
<span class="nf">otu_table</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">)</span> <span class="o">=</span> <span class="nf">otu_table</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">)</span><span class="m">+1</span>
<span class="c1"># make deseq object from phyloseq object</span>
<span class="n">ds</span> <span class="o">=</span> <span class="nf">phyloseq_to_deseq2</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">,</span> <span class="o">~</span> <span class="n">Group1</span><span class="p">)</span>
<span class="c1"># Run DESeq2</span>
<span class="n">dds.data</span> <span class="o">=</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="c1"># With the &#39;contrast&#39; function you screen two different set of samples (based on your metadata) for differential taxa.</span>
<span class="n">res</span> <span class="o">=</span> <span class="nf">results</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span><span class="s">&quot;Dental_calculus&quot;</span><span class="p">,</span><span class="s">&quot;Oral_plaque&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>Then you can explore your results, filter and sort the differential taxa detected based on a False Dicovery Rate (FDR) threshold (e.g. set to 0.01), the log2-fold-change and the base mean. Read more about the <a class="reference external" href="https://www.biostars.org/p/209118/">FDR threshold here</a>.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort based on p-value adjusted:</span>
<span class="n">resOrdered</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">,</span> <span class="n">na.last</span><span class="o">=</span><span class="kc">NA</span><span class="p">),</span> <span class="p">]</span>
<span class="c1"># set a threshold value for the False Discovery Rate (FDR):</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="m">0.01</span>
<span class="c1"># get only significant taxa based on p-value adjusted (the FDR):</span>
<span class="n">resSig</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">resOrdered</span><span class="p">,</span> <span class="n">padj</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">)</span>
<span class="c1"># sort significant values based on the log2-fold-change:</span>
<span class="n">resfc</span> <span class="o">=</span> <span class="n">resSig</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">resSig</span><span class="o">$</span><span class="n">log2FoldChange</span><span class="p">),]</span>
<span class="c1"># sort significant values based on abundance (the base mean):</span>
<span class="n">resbm</span> <span class="o">=</span> <span class="n">resSig</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">resSig</span><span class="o">$</span><span class="n">baseMean</span><span class="p">),]</span>
<span class="c1"># save the the tables</span>
<span class="nf">write.table</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">resbm</span><span class="p">),</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;deseq2.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">col.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A positive log2-fold-change for a comparison of A vs B means that the feature (OTU) in A is more abundant than in B (and viceversa).
For example, a log2-fold-change of −1 means that in A the OTU is of 2^−1 = 0.5 less abundant than in B.</p>
</div>
<p>You can plot the log2-fold-changes to visualize for example the differential distribution of species based on their phyla (or genera) in the two groups contrasted. You could also color each differential taxon based on the phylum it belongs to.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># convert the data in a dataframe</span>
<span class="n">resSigMod</span> <span class="o">=</span> <span class="nf">cbind</span><span class="p">(</span><span class="nf">as</span><span class="p">(</span><span class="n">resSig</span><span class="p">,</span> <span class="s">&quot;data.frame&quot;</span><span class="p">),</span> <span class="nf">as</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">)[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">resSig</span><span class="p">),</span> <span class="p">],</span> <span class="s">&quot;matrix&quot;</span><span class="p">))</span>
<span class="c1"># Plot log-fold-changes of the OTUs based on Phylum</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">resSigMod</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">phylum</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">log2FoldChange</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">phylum</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_jitter</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">vjust</span><span class="o">=</span><span class="m">0.5</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Dental calculus vs Oral plaque&quot;</span><span class="p">)</span>
<span class="c1"># Plot log-fold-changes of the OTUs based on genera</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">resSigMod</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">genus</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">log2FoldChange</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">phylum</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_jitter</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">vjust</span><span class="o">=</span><span class="m">0.5</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Dental calculus vs Oral plaque&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also generate a heatmap of the transformed abundance data. For visualization or clustering purposes it might be useful to work with transformed versions of the count data.
Various <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization">transformations</a> may be applied on the data:
1) ntd: Log transformation of the the data
2) vsd: Variance Stabilizing Transformation
3) rld: Regularized log transformation</p>
<blockquote>
<div></div></blockquote>
<p>When dealing with many samples it is useful to use the vst. To make the various transformatinos run the folowing commands:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># variance stabilizing transformation</span>
<span class="n">vsd</span> <span class="o">=</span> <span class="nf">varianceStabilizingTransformation</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">blind</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="c1"># log transformation</span>
<span class="n">ntd</span> <span class="o">&lt;-</span> <span class="nf">normTransform</span><span class="p">(</span><span class="n">dds.data</span><span class="p">)</span>
<span class="c1"># regularized log transformation</span>
<span class="n">rld</span> <span class="o">&lt;-</span> <span class="nf">rlog</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">blind</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<p>We will plot the heatmap of the transformed data only for a reduced number of taxa (with decreasing counts).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;pheatmap&quot;</span><span class="p">)</span>
<span class="c1">#adjust max number of taxa to display.</span>
<span class="n">select</span> <span class="o">&lt;-</span> <span class="nf">order</span><span class="p">(</span><span class="nf">rowMeans</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span><span class="n">normalized</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span>
                <span class="n">decreasing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">]</span>
<span class="c1"># create a dataframe importing using the metadata (e.g. Group1) and incorporating the taxa names as row names.</span>
<span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">colData</span><span class="p">(</span><span class="n">dds.data</span><span class="p">)[,</span><span class="s">&quot;Group1&quot;</span><span class="p">])</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">subsample_flt</span><span class="o">@</span><span class="n">otu_table</span><span class="p">)</span>
<span class="c1"># plot heatmap (on ntd transformation)</span>
<span class="nf">pheatmap</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">vsd</span><span class="p">)[</span><span class="n">select</span><span class="p">,],</span> <span class="n">cluster_rows</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> <span class="n">show_rownames</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span>
         <span class="n">cluster_cols</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> <span class="n">annotation_col</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="m">4</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="principal-component-analysis-pca">
<h2><span class="section-number">5.9. </span>Principal Component Analysis (PCA)<a class="headerlink" href="#principal-component-analysis-pca" title="Permalink to this headline"></a></h2>
<p>The Principal Component Analysis (PCA) is another multivariate reduction method used for data visualization.
It does not use a distance matrix and the benefit of that is that you can plot loadings onto your PCA axes.
You can read more about the general principles of PCA here: <a class="reference external" href="https://builtin.com/data-science/step-step-explanation-principal-component-analysis">link3</a>, <a class="reference external" href="https://archetypalecology.wordpress.com/2018/02/17/principal-component-analysis-in-r/">link4</a></p>
<blockquote>
<div></div></blockquote>
<p>We will make the PCA only on the oral microbiomes (dental calculus and plaque). So we will use the subsampled table that we already used for <a class="reference internal" href="#deseq2"><span class="std std-ref">DESeq2</span></a>.
Make sure you have the subsampled table, or do that again as follows:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample oral microbiomes</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Dental_calculus_chimp&quot;</span><span class="p">,</span><span class="s">&quot;Dental_calculus&quot;</span><span class="p">,</span><span class="s">&quot;Oral_plaque&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">Group1</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1">#filter out low-abundance taxa with a mean &lt; 0.02%.</span>
<span class="n">minTotRelAbun</span> <span class="o">=</span> <span class="m">0.0002</span>
<span class="n">x</span> <span class="o">=</span> <span class="nf">taxa_sums</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span>
<span class="n">keepTaxa</span> <span class="o">=</span> <span class="nf">taxa_names</span><span class="p">(</span><span class="n">subsample</span><span class="p">)[</span><span class="nf">which</span><span class="p">((</span><span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">minTotRelAbun</span><span class="p">)]</span>
<span class="n">subsample_flt</span> <span class="o">=</span> <span class="nf">prune_taxa</span><span class="p">(</span><span class="n">keepTaxa</span><span class="p">,</span> <span class="n">subsample</span><span class="p">)</span>
<span class="c1"># Convert to relative frequencies.</span>
<span class="n">subsample_rel</span> <span class="o">=</span> <span class="nf">transform_sample_counts</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is better to remove the low-abundant taxa only <strong>after</strong> subsampling the abudance table. Doing this before may keep unwanted taxa that are more common in other microbiota (e.g. soil) and irrelevant for describing oral envrionments (or the specific environment you are analysing with the PCA).</p>
</div>
<p>We will do the clr-transformation of the otu table in the biom object with the package <cite>microbiome</cite>, that we already <a class="reference internal" href="#installed-before"><span class="std std-ref">installed before</span></a>
Activate the library and transform the data with the following command:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">microbiome</span><span class="p">)</span>
<span class="n">subsample_rel_clr</span> <span class="o">=</span> <span class="nf">transform</span><span class="p">(</span><span class="n">subsample_rel</span><span class="p">,</span><span class="s">&quot;clr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, make the PCA with the command <cite>ordinate</cite>, adopting the ordination method <cite>RDA</cite> (redundancy analysis, a constrained ordination method), which without constraints correspond to the PCA in phyloseq.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ord_clr</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample_rel_clr</span><span class="p">,</span> <span class="s">&quot;RDA&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can plot the variance associated with each PC:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_scree</span><span class="p">(</span><span class="n">ord_clr</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;blue&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;\nAxis&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Proportion of Variance\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>And finally plot the PCA in a 2D chart.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample_rel_clr</span><span class="p">,</span> <span class="n">ord_clr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
<p>To display the labels of each symbol:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample_rel_clr</span><span class="p">,</span> <span class="n">ord_clr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s">&quot;Sample_short&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To refine the analysis you can remove unwanted samples (e.g. those that appeared to be contaminated or outliers). Let’s remove these four samples in the oral plaque group and repeat the PCA.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">remove</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;SRR1564193&quot;</span><span class="p">,</span><span class="s">&quot;SRR1564223&quot;</span><span class="p">,</span><span class="s">&quot;SRR1564208&quot;</span><span class="p">,</span><span class="s">&quot;SRR1564211&quot;</span><span class="p">)</span>
<span class="n">subsample_pruned</span> <span class="o">=</span> <span class="nf">prune_samples</span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="n">subsample_rel_clr</span><span class="o">@</span><span class="n">sam_data</span><span class="o">$</span><span class="n">Sample_short</span> <span class="o">%in%</span> <span class="n">remove</span><span class="p">),</span> <span class="n">subsample_rel_clr</span><span class="p">)</span>
<span class="n">ord_clr_pruned</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample_pruned</span><span class="p">,</span> <span class="s">&quot;RDA&quot;</span><span class="p">)</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample_pruned</span><span class="p">,</span> <span class="n">ord_clr_pruned</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group1&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="4_prepare_tables.html" class="btn btn-neutral float-left" title="4. Prepare abundance tables for downstream analyses" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="6_Do_it_yourself.html" class="btn btn-neutral float-right" title="6. Do it yourself" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Claudio Ottoni.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>